### DISCUSSION TITLE
Adding more fine-grained logic for repo mirroring

### DISCUSSION THREAD

**Discussion by brhenc** [quay/quay#3120] [General]
Currently quay retries an entire run if a single tag fails to be mirrored using skopeo. 

Would it be possible to retry an individual tag a configurable number of times (e.g. 3) before retrying the entire run?

The use case is to mirror tagged point releases (about ~200 releases per repository) from an internal quay registry 
to a public-facing quay registry. If a single transfer fails for any reason (temporary network failure) the entire run is retried.

Additionally, it would be useful to have the option to mirror only the tags that have changed between the two quay registries. This would reduce the number of tags that are part of a run and reduce the likelihood a run will fail.

Mostly looking for pointers on how to approach making contributions in this space (mirroring) and to check if a PR would be accepted if I made one, or if somebody is already working on something similar.

Thank you in advance.

**Comment by ibazulic** [comment]
Hi @brhenc ,

Quay mirrors do not mirror data that has already been mirroed. However, because we're working on tags that are of a floating nature, it's important to check them all during mirroring so that new versions of images are properly pulled. 

As far as mirroring itself goes, the UI should have an option to ignore failures, and preserve everything that was mirrored up to that point. If a tag wasn't mirrored this time around, it will be mirrored the 2nd time when the cycle starts again. 

Does that answer your question?

**Reply by brhenc** [reply]
I haven't been able to add a label, the question was answered during the quay community meeting.

The suggestion was (roughly) to:
1)  Try using the proxy feature
2) Try using skopeo directly as part of a CI/CD pipeline. 

I have to follow-up on the above options.

Regarding your suggestions:
Mirroring existing tags isn't a good thing in some contexts. In a sense, tags that are visible to customers are part of a "stable" API and become the source of truth. Accidentally mirroring an existing tag that was modified is undesirable in case of malicious/nefarious actors. 

On ignoring failures - I will check, but that wasn't the behavior for the quay instance I have. It helps to know that 
there might be a combination of options that does behave differently.


The preliminary conclusion might be that for a CI/CD pipeline, it's best to use skopeo and see if skopeo can support such a (custom) workflow. If it can't, it's a limitation in skopeo and not in quay.

Thank you very much for the fast reply @ibazulic and all the suggestions! I have quite a few things to follow-up on.

It might take me some time to respond/follow-up but I think I have all the answers I need.
